firewall_test.go

```diff
@@ -304,3 +304,73 @@ func TestProviders(t *testing.T) {
 		}
 	}
 }
+
+func TestFilter(t *testing.T) {
+	// Create a new dynamic session
+	s, err := New(&Options{
+		Dynamic: true,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer s.Close()
+
+	guid, err := windows.GenerateGUID()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a new provider
+	p := &Provider{
+		ID:          ProviderID(guid),
+		Name:        "test provider",
+		Description: "a test provider",
+		Data:        []byte("byte blob"),
+	}
+	if err := s.AddProvider(p); err != nil {
+		t.Fatalf("add provider failed: %v", err)
+	}
+
+	rule_id, err := windows.GenerateGUID()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a rule so we can test enumeration
+	r := Rule{
+		ID:          RuleID(rule_id),
+		Name:        "inetaf-wf-test-rule",
+		Description: "A disabled rule for testing the inet.af/wf library",
+		Layer:       LayerALEAuthConnectV4,
+		Weight:      0x4242,
+		Conditions: []*Match{
+			{
+				Field: FieldALEAppID,
+				Op:    MatchTypeEqual,
+				Value: "test",
+			},
+		},
+		Action:     ActionPermit,
+		Persistent: false,
+		BootTime:   false,
+		Provider:   ProviderID(guid),
+		Disabled:   true,
+	}
+	if err := s.AddRule(&r); err != nil {
+		t.Fatal(err)
+	}
+
+	// Filter for rules from our provider
+	rules, err := s.EnumerateRules(FilterEnumTypeOverlapping, LayerALEAuthConnectV4).
+		WithProvider(ProviderID(guid)).
+		WithActionMask(0xFFFFFFFF).
```

Comment:
In case it's worth the effort now... if not, we can iterate depending on the consumption in the future.

For the action mask I was thinking defining it in terms of a type vs. the raw uint32...
Technically `wf` has the `Action` "enum" but it's not exactly listed like the Win SDK has it since a few combinations enum flag combos were already done.  

The thought of unwinding this as part of this PR makes me feel like it's bigger in scope, which is my I'm meh on whether to do this or not, and will defer to you.
```
#define FWP_ACTION_FLAG_TERMINATING     (0x00001000)
#define FWP_ACTION_FLAG_NON_TERMINATING (0x00002000)
#define FWP_ACTION_FLAG_CALLOUT         (0x00004000)
typedef UINT32 FWP_ACTION_TYPE;

#define FWP_ACTION_BLOCK  (0x00000001 | FWP_ACTION_FLAG_TERMINATING)
#define FWP_ACTION_PERMIT (0x00000002 | FWP_ACTION_FLAG_TERMINATING)
#define FWP_ACTION_CALLOUT_TERMINATING (0x00000003 | FWP_ACTION_FLAG_CALLOUT | FWP_ACTION_FLAG_TERMINATING)
#define FWP_ACTION_CALLOUT_INSPECTION (0x00000004 | FWP_ACTION_FLAG_CALLOUT | FWP_ACTION_FLAG_NON_TERMINATING)
#define FWP_ACTION_CALLOUT_UNKNOWN (0x00000005 | FWP_ACTION_FLAG_CALLOUT)
#define FWP_ACTION_CONTINUE (0x00000006 | FWP_ACTION_FLAG_NON_TERMINATING)
#define FWP_ACTION_NONE (0x00000007)
#define FWP_ACTION_NONE_NO_MATCH (0x00000008)
```

Reply:
Are you fine with this?

```go
// Action is an action the filtering engine can execute.
type Action uint32

const (
	// Action Flags can be used in enumerations to restrict the type of rules to return
	ActionFlagTerminating    Action = 0x1000     // only rules which must terminate evaluation
	ActionFlagNonTerminating        = 0x2000     // only rules which cannot terminate evaluation
	ActionFlagCallout               = 0x4000     // any rules with a callout
	ActionFlagIgnore                = 0xFFFFFFFF // match any rule action

	// pre-defined actions with appropriate action flags
	ActionBlock              Action = (0x01 | ActionFlagTerminating)                        // blocks a packet or session
	ActionPermit                    = (0x02 | ActionFlagTerminating)                        // permits a packet or session
	ActionCalloutTerminating        = (0x03 | ActionFlagCallout | ActionFlagTerminating)    // invokes a callout for permit/block
	ActionCalloutInspection         = (0x04 | ActionFlagCallout | ActionFlagNonTerminating) // invokes a callout for inspection only
	ActionCalloutUnknown            = (0x05 | ActionFlagCallout)                            // invokes a callout which may return permit or block
)
```